jsonrpc/debug_endpoint.go

```diff
@@ -0,0 +1,209 @@
+
+	// GetBlockByNumber gets a block using the provided height
+	GetBlockByNumber(num uint64, full bool) (*types.Block, bool)
+
+	TraceMinedBlock(*types.Block, tracer.Tracer) ([]interface{}, error)
+
+	TraceMinedTxn(*types.Block, types.Hash, tracer.Tracer) (interface{}, error)
+
+	TraceCall(*types.Transaction, *types.Header, tracer.Tracer) (interface{}, error)
+}
+
+type debugTxPoolStore interface {
+	GetNonce(types.Address) uint64
+}
+
+type debugStateStore interface {
+	GetAccount(root types.Hash, addr types.Address) (*state.Account, error)
+}
+
+type debugStore interface {
+	debugBlockchainStore
+	debugTxPoolStore
+	debugStateStore
+}
+
+// Debug is the debug jsonrpc endpoint
+type Debug struct {
+	store debugStore
+}
+
+type TraceConfig struct {
+	EnableMemory     bool `json:"enableMemory"`
+	DisableStack     bool `json:"disableStack"`
+	DisableStorage   bool `json:"disableStorage"`
+	EnableReturnData bool `json:"enableReturnData"`
+	// Timeout *string `json:"timeout"`
+}
+
+func (d *Debug) TraceBlockByNumber(
+	blockNumber BlockNumber,
+	config *TraceConfig,
+) (interface{}, error) {
+	num, err := GetNumericBlockNumber(blockNumber, d.store)
+	if err != nil {
+		return nil, err
+	}
+
+	block, ok := d.store.GetBlockByNumber(num, true)
+	if !ok {
+		return nil, errors.New("block not found")
+	}
+
+	return d.traceBlock(block, config)
+}
+
+func (d *Debug) TraceBlockByHash(
+	blockHash types.Hash,
+	config *TraceConfig,
+) (interface{}, error) {
+	block, ok := d.store.GetBlockByHash(blockHash, true)
+	if !ok {
+		return nil, errors.New("block not found")
+	}
+
+	return d.traceBlock(block, config)
+}
+
+func (d *Debug) TraceBlock(
+	input string,
+	config *TraceConfig,
+) (interface{}, error) {
+	blockByte, decodeErr := hex.DecodeHex(input)
+	if decodeErr != nil {
+		return nil, fmt.Errorf("unable to decode block, %w", decodeErr)
+	}
+
+	block := &types.Block{}
+	if err := block.UnmarshalRLP(blockByte); err != nil {
+		return nil, err
+	}
+
+	return d.traceBlock(block, config)
+}
+
+func (d *Debug) TraceTransaction(
+	txHash types.Hash,
+	config *TraceConfig,
+) (interface{}, error) {
+	tx, block := GetTxAndBlockByTxHash(txHash, d.store)
+	if tx == nil {
+		// TODO: check error schema
+		return nil, fmt.Errorf("tx %s not found", txHash.String())
+	}
+
+	if block.Number() == 0 {
+		return nil, errors.New("genesis is not traceable")
+	}
+
+	tracer := newTracer(config)
+
+	return d.store.TraceMinedTxn(block, tx.Hash, tracer)
+}
+
+func (d *Debug) TraceCall(
+	arg *txnArgs,
+	filter BlockNumberOrHash,
+	config *TraceConfig,
+) (interface{}, error) {
+	var (
+		header *types.Header
+		err    error
+	)
+
+	// The filter is empty, use the latest block by default
+	if filter.BlockNumber == nil && filter.BlockHash == nil {
+		filter.BlockNumber, _ = createBlockNumberPointer("latest")
+	}
+
+	header, err = d.getHeaderFromBlockNumberOrHash(&filter)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get header from block hash or block number")
```

Comment:
do we need fmt.Errorf here? maybe errors.New or fmt.Errorf + come additional data?

Reply:
Fixed 1c181d3f8ebdb9db2d0742371ef85c814307eb52